main/src/test/java/org/mobilitydata/gtfsvalidator/util/CalendarUtilTest.java

```diff
@@ -0,0 +1,204 @@
+        assertThat(servicePeriod.getAddedDays())
+                .containsExactly(LocalDate.of(2021, 1, 6));
+        assertThat(servicePeriod.getRemovedDays())
+                .containsExactly(LocalDate.of(2021, 1, 7));
+    }
+
+    @Test
+    public void CreateServicePeriodEmpty() {
+        ServicePeriod servicePeriod =
+                CalendarUtil.createServicePeriod(null, ImmutableList.of());
+        assertThat(servicePeriod.getServiceStart()).isEqualTo(LocalDate.EPOCH);
+        assertThat(servicePeriod.getServiceEnd()).isEqualTo(LocalDate.EPOCH);
+        assertThat(servicePeriod.getWeeklyPattern()).isEqualTo(0);
+        assertThat(servicePeriod.getAddedDays()).isEmpty();
+        assertThat(servicePeriod.getRemovedDays()).isEmpty();
+    }
+
+    @Test
+    public void buildServicePeriodMap() {
+        NoticeContainer noticeContainer = new NoticeContainer();
+        GtfsCalendarTableContainer calendarTable =
+                GtfsCalendarTableContainer.forEntities(
+                        ImmutableList.of(createGtfsCalendar(
+                                "s1", LocalDate.of(2021, 1, 4), LocalDate.of(2021, 1, 5),
+                                ImmutableSet.of(DayOfWeek.MONDAY, DayOfWeek.TUESDAY,
+                                        DayOfWeek.WEDNESDAY, DayOfWeek.FRIDAY)),
+                                createGtfsCalendar(
+                                        "s2", LocalDate.of(2021, 2, 14), LocalDate.of(2021, 2, 15),
+                                        ImmutableSet.of(DayOfWeek.SUNDAY))),
+                        noticeContainer);
+
+        GtfsCalendarDateTableContainer calendarDateTable =
+                GtfsCalendarDateTableContainer.forEntities(
+                        ImmutableList.of(
+                                addedCalendarDate("s1", LocalDate.of(2021, 1, 6)),
+                                removedCalendarDate("s1", LocalDate.of(2021, 1, 7)),
+                                addedCalendarDate("s3", LocalDate.of(2021, 3, 8)),
+                                removedCalendarDate("s3", LocalDate.of(2021, 3, 9))),
+                        noticeContainer);
+
+        assertThat(
+                CalendarUtil.buildServicePeriodMap(calendarTable, calendarDateTable))
+                .containsExactly(
+                        "s1",
+                        new ServicePeriod(LocalDate.of(2021, 1, 4),
+                                LocalDate.of(2021, 1, 5), (byte) 0b0010111,
+                                ImmutableSet.of(LocalDate.of(2021, 1, 6)),
+                                ImmutableSet.of(LocalDate.of(2021, 1, 7))),
+                        "s2",
+                        new ServicePeriod(LocalDate.of(2021, 2, 14),
+                                LocalDate.of(2021, 2, 15), (byte) 0b1000000,
+                                ImmutableSet.of(), ImmutableSet.of()),
+                        "s3",
+                        new ServicePeriod(LocalDate.EPOCH, LocalDate.EPOCH, (byte) 0,
```

Comment:
I would like to share an principle formulated by a Googler that I usually follow. What do you think?

> Put only one CL comment for each high-level. If you have a single logical suggestion for a change to the CL that applies to multiple lines, avoid adding a review comment to every single line to which it applies.  Instead, put a single comment in the most logical line (which is frequently the first line to which it applies) and add an "Also elsewhere." (or "Also below." or "Please fix throughout.") clause at the end.
> Rationale
> It may be tempting to flag explicitly all the lines that need to be changed under the illusion that, since you'll need to verify that the fix has been applied throughout, you might as well go over all the code now and make sure you capture the list of lines that need changes (by commenting on each), to save the code author some time.  However, that rarely ends up saving time: the code author now has to match all your comments with his code and click "Done" many times.  Frequently, it's more efficient to just let the author clean up the code until he's reasonably sure he's applied the comment throughout.

Reply:
☝️ SGTM!